mesa/agent.py

```diff
@@ -20,19 +20,35 @@
 
 
 class Agent:
-    """Base class for a model agent."""
+    """
+    Base class for a model agent in Mesa.
+
+    Attributes:
+        unique_id (int): A unique identifier for this agent.
+        model (Model): A reference to the model instance.
+        self.pos: Position | None = None
+    """
 
     def __init__(self, unique_id: int, model: Model) -> None:
-        """Create a new agent.
+        """
+        Create a new agent.
 
         Args:
-            unique_id (int): A unique numeric identified for the agent
-            model: (Model): Instance of the model that contains the agent
+            unique_id (int): A unique identifier for this agent.
+            model (Model): The model instance in which the agent exists.
         """
         self.unique_id = unique_id
         self.model = model
         self.pos: Position | None = None
 
+        # Register the agent with the model using defaultdict
+        self.model.agents[type(self)].add(self)
+
+    def remove(self) -> None:
+        """Remove and delete the agent from the model."""
+        self.model.agents[type(self)].discard(self)
+        del self
```

Comment:
2 things:

1. del is not needed here
2. this does make it important to get the weakref stuff in schedulers and spaces otherwise you get a memory leak or the user will have to explicly extend remove.

Reply:
Yes, agreed. [`weakref`](https://docs.python.org/3/library/weakref.html) is a really nice solution here since currently it's a lot more complex to delete agents, also because you can't assume there's exactly one scheduler or grid, both can be 0 or more.

Reply:
Right, it's because it only deletes the reference and not the actual object. The object is deleted in Python when its reference count drops to zero, and is then removed by the garbage collector.